python/pyspark/graphx/vertex.py

```diff
@@ -0,0 +1,330 @@
+        target_storage_level = StorageLevel.MEMORY_ONLY
+        java_storage_level = self.ctx._getJavaStorageLevel(target_storage_level)
+        prdd = self.ctx._jvm.PythonVertexRDD(rdd._jrdd,
+                                                   bytearray(pickled_command),
+                                                   env, includes, self.preserve_partitioning,
+                                                   self.ctx.pythonExec,
+                                                   broadcast_vars, self.ctx._javaAccumulator,
+                                                   java_storage_level)
+        self.jvertex_rdd = prdd.asJavaVertexRDD()
+        if enable_profile:
+            self.id = self.jvertex_rdd.id()
+            self.ctx._add_profile(self.id, profileStats)
+        return self.jvertex_rdd
+
+
+class PipelinedVertexRDD(VertexRDD):
+
+    """
+    Pipelined mapValues in VertexRDD:
+
+    >>> rdd = sc.parallelize([(1, ("Alice", 29)), (2, ("Bob", 30)), \
+                              (3, ("Charlie", 31)), (4, ("Dwayne", 32))])
+    >>> vertices = VertexRDD(rdd)
+    >>> vertices.mapValues(lambda x: x[1] * 2).cache().collect()
+    [(1, ("Alice", 58)), (2, ("Bob", 60)), \
+     (3, ("Charlie", 62)), (4, ("Dwayne", 64))]
+
+    Pipelined reduces in VertexRDD:
+    >>> from operator import add
+    >>> rdd.map(lambda x: 2 * x).reduce(add)
+    20
+    >>> rdd.flatMap(lambda x: [x, x]).reduce(add)
+    20
+    """
+
+    def __init__(self, prev, func, preservesPartitioning=False):
+        if not isinstance(prev, PipelinedVertexRDD) or not prev.is_pipelinable():
+            # This transformation is the first in its stage:
+            self.func = func
+            self.preservesPartitioning = preservesPartitioning
+            self.prev_jvertex_rdd = prev.jvertex_rdd
+            self.prev_jvertex_rdd_deserializer = prev.jvertex_rdd_deserializer
+        else:
+            prev_func = prev.func
+
+            def pipeline_func(split, iterator):
+                return func(split, prev_func(split, iterator))
+            self.func = pipeline_func
+            self.preservesPartitioning = \
+                prev.preservesPartitioning and preservesPartitioning
+            self.prev_jvertex_rdd = prev.jvertex_rdd
+            self.prev_jvertex_rdd_deserializer = prev.prev_jvertex_rdd_deserializer
+
+        self.is_cached = False
+        self.is_checkpointed = False
+        self.ctx = prev.ctx
+        self.prev = prev
+        self.jvrdd_val = None
+        self.id = None
+        self.jvertex_rdd_deserializer = self.ctx.serializer
+        self.bypass_serializer = False
+        self.partitionFunc = prev._partitionFunc if self.preservesPartitioning else None
+        self.broadcast = None
+
+    def __del__(self):
+        if self.broadcast:
+            self.broadcast.unpersist()
+            self.broadcast = None
+
+    @property
+    def jvertex_rdd(self):
```

Comment:
There's a lot of duplication between this method, the one in edge RDD, and the regular PySpark RDD.  I'm worried that this could become a maintenance problem since lots of common code is duplicated and will fall out of sync.  Do you know if there's a good way to add some abstraction here so that we're not copying, say, the profiling code into three places?